dev-tools/mage/pkg.go

```diff
@@ -0,0 +1,108 @@
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package mage
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/magefile/mage/mg"
+	"github.com/magefile/mage/sh"
+	"github.com/pkg/errors"
+)
+
+// Package packages the Beat for distribution. It generates packages based on
+// the set of target plaforms and registered packaging specifications.
+func Package() error {
+	if len(Platforms) == 0 {
+		return errors.New("PLATFORMS environment must be set to a list of " +
+			"GOOS/Arch values, but Platforms is empty")
+	}
+
+	if len(Packages) == 0 {
+		return errors.New("no package specs are registered. Call " +
+			"UseCommunityBeatPackaging or UseElasticBeatPackaging first.")
+	}
+
+	var tasks []interface{}
+	for _, target := range Platforms {
+		for _, pkg := range Packages {
+			if pkg.OS != target.GOOS() {
+				continue
+			}
+
+			for _, pkgType := range pkg.Types {
+				packageArch, err := getOSArchName(target, pkgType)
+				if err != nil {
+					log.Printf("Skipping arch %v for package type %v: %v", target.Arch(), pkgType, err)
+					continue
+				}
+
+				spec := pkg.Spec.Clone()
+				spec.OS = target.GOOS()
+				spec.Arch = packageArch
+				spec.Snapshot = Snapshot
+				spec.evalContext = map[string]interface{}{
+					"GOOS":        target.GOOS(),
+					"GOARCH":      target.GOARCH(),
+					"GOARM":       target.GOARM(),
+					"Platform":    target,
+					"PackageType": pkgType.String(),
+					"BinaryExt":   binaryExtension(target.GOOS()),
+				}
+				spec.packageDir = packageStagingDir + "/" + pkgType.AddFileExtension(spec.Name+"-"+target.GOOS()+"-"+target.Arch())
+				spec = spec.Evaluate()
+
+				tasks = append(tasks, packageBuilder{target, spec, pkgType}.Build)
+			}
+		}
+	}
+
+	Parallel(tasks...)
+	return nil
+}
+
+type packageBuilder struct {
+	Platform BuildPlatform
+	Spec     PackageSpec
+	Type     PackageType
+}
+
+func (b packageBuilder) Build() error {
+	fmt.Printf(">> package: Building %v type=%v for platform=%v\n", b.Spec.Name, b.Type, b.Platform.Name)
+	log.Printf("Package spec: %+v", b.Spec)
+	return errors.Wrapf(b.Type.Build(b.Spec), "failed building %v type=%v for platform=%v",
+		b.Spec.Name, b.Type, b.Platform.Name)
+}
+
+// TestPackages executes the package tests on the produced binaries. These tests
```

Comment:
Really nice. We could also extend these with things like checking if the dashboard files are all in the correct directory etc. ¨

Reply:
Good idea. I suggest enhancing the test outside of this PR. The package_test.go is executed by the release-manager today.